### PR TITLE
Skeleton

### PR DESCRIPTION
**Pull request by kris-lab**
----
Still to do
- [x] error handling for Janus messages
- [x] error handling for plugins messages
- [x] concurrent/promise error handling
- [x] event-machine error handling
- [x] more specs

### CODE REVIEW

**Comment by kris-lab** [comment]
@njam this is not ready but:
- could you have a look on `janus/plugin/rtpbroadcast/mountpoint` structure - do you think it is good idea to introduce plugin/features based classes?
- should we have real test with travis? I mean to install janus + plugins and then run specs? see current spec for `mountpoint`?

**Comment by njam** [comment]
- session, plugin, mountpoint: How about introducing a common *parent* like `Janus::Resource` (which would always keep the reference to the client, for sending commands)
 - yes, agree to have this plugin-specific resource
- client: Only expose attributes that are needed (probably client?)
- client: Rename "client" to "websocket_client"?
- Regarding tests: I think it would be better to mock the websocket server in this case. Tests will be much faster, and the server logic is not so complex that it's too big of a difference. wdyt?

**Comment by kris-lab** [comment]
So you mean:
- `Janus::Resource::Session`
- `Janus::Resource::Plugin`
- `Janus::Resource::Plugin::Rtpbroadcast::Mountpoint`

or rather

- `Janus::Resource::Session`
- `Janus::Resource::Plugin`
- `Janus::Plugin::Rtpbroadcast::Mountpoint`

?

**Comment by njam** [comment]
I would go with the latter (keeping all files of this plugin together), but extending from `Janus::Resource`, wdyt?

**Comment by kris-lab** [comment]
@najm if you have sec please have a look on current structure
- there are `resources` for janus
- there are `resources` for plugins
- there is very basic mock for websocket client (it must be moved out to some spec_helper?)

I will finish and polish it up if you agree on the structure. There are some missing parts still like maybe callback for janus errors which could be automatically detected by `client` from incoming messages.

**Comment by njam** [comment]
briefly checked (need to leave, sorry):
- resource: I would add the `client` getter here. Maybe even a constructor which expects a client instance? Then all child classes would access the `client` from the common getter, and it would be in the structure of the classes that they need to have it.
- resource: maybe include EventEmitter? And use common events, like `create` and `destroy`?
- Plugin::Rtpbroadcast::Resource: I think it's not needed?
- General: I would drop all attr_accessors that are not needed (not public interface). Let's only expose public attributes, and public methods that are useful. The rest can be made private.

**Comment by kris-lab** [comment]
@njam thanks for tips. agree on all

**Comment by kris-lab** [comment]
@njam fyi https://github.com/sqwiggle/janus-gateway-ruby

**Comment by njam** [comment]
ah sqwiggle… :D
Thought about that name too, would be more expected. Should we rename to "janus-gateway-ruby"?

**Comment by kris-lab** [comment]
I think we should rename!

**Comment by njam** [comment]
Changed the repo name, can you adjust the code?

**Comment by kris-lab** [comment]
@njam so how should structure this
- modules
```ruby
module Janus
 module Gateway
  class Client
  end
 end
end
```

or rather
```ruby
module JanusGateway
 class Client
 end
end
```

- `gemspec`
```ruby
s.name        = 'janus-gateway-ruby'
```
or
```ruby
s.name        = 'janus-gateway_ruby'
```

Or just update `gemspec` and keep stuructre like it is now?

**Comment by kris-lab** [comment]
@njam please review

**Comment by kris-lab** [comment]
Two things:
- naming is not adjusted but we have to decide in it
- there should be layer of transport for this client. currently we have websocket integrated quite deeply but there should be easy option to extend with `http` or `rabbit` if any contributor is willing to do that! wdyt?

**Comment by njam** [comment]
Because "janus-gateway" is *one* term and we don't want to reflect "ruby" in the gem's name let's do:
```ruby
module JanusGateway
 class Client
 end
end
```

Then maybe as a gem name `janus_gateway`?

**Comment by njam** [comment]
Agree on the transports topic. I guess sending and receiving data from the server is only happening in `JanusGateway::Client`?

**Comment by kris-lab** [comment]
- yes for naming
- yes for transport

**Comment by kris-lab** [comment]
@njam renamed! what about transport? should I implement it now?

**Comment by njam** [comment]
Yes, let's add transport, makes sense and should be simple. Could also help for testing (we can stub the whole transport).

**Comment by kris-lab** [comment]
@njam ok, I have implemented the transport layer for webscoket! have a look. It is a bit rough but should give a lot of flexibility!?

**Comment by njam** [comment]
- janus_gateway/client.rb: I would drop the auto transport generation in favor of a clean signature (only transport, no url). wdyt? -> adjust readme
- janus_gateway/client.rb: Remove faye import
- client_spec.rb: Probably move to transport spec?
- transports: Adapter pattern looks good!

Could you add docu everywhere, will make it easier to review!

**Comment by kris-lab** [comment]
@njam please have a look now

**Comment by njam** [comment]
- README: Rtpbrodcast: Last code example: Move `.then` on new line, so there's only one level of indentation?
 - Why do we need `_self.client` - is there another `client` on the `Client`?
- plugin/rtpbroadcast.rb: How about making this extend the base Plugin? Then we could do `JanusGateway::Plugin::Rtpbroadcast.new()` (for example in the readme!)
- plugin/rtpbroadcast/resource.rb: How about making this extend the base Resource?
- transport.rb: initialize: Add docu
- transport.rb: send: Fix docu
- transport.rb: ready_state: What are possible return values? Should we document them?
- transport.rb: has_client? / has_connection?: On an abstract level I suggest we don't care about a "client". How about combining those two into `is_connected?`?- transport/websocket.rb: connect: 
- transport/websocket.rb: connect: I would suggest to not pass on `event` when emitting `open`. We don't need it and it could be confusing wdyt?
-

**Comment by njam** [comment]
- README: Rtpbrodcast: Last code example: Move `.then` on new line, so there's only one level of indentation?
 - Why do we need `_self.client` - is there another `client` on the `Client`?
- plugin/rtpbroadcast.rb: How about making this extend the base Plugin? Then we could do `JanusGateway::Plugin::Rtpbroadcast.new()` (for example in the readme!)
- plugin/rtpbroadcast/resource.rb: How about making this extend the base Resource?
- transport.rb: initialize: Add docu
- transport.rb: send: Fix docu
- transport.rb: ready_state: What are possible return values? Should we document them?
- transport.rb: has_client? / has_connection?: On an abstract level I suggest we don't care about a "client". How about combining those two into `is_connected?`?
- transport/websocket.rb: connect:  In the error handler we now raise, where will this be rescued? Should it eventually be logged? Maybe add a todo there to remember?
- transport/websocket.rb: connect: I would suggest to not pass on `event` when emitting `open`. We don't need it and it could be confusing wdyt?
- transport/websocket.rb: connect: The transaction_queue is a Hash where key is the transaction's id right? So we can just get that one promise if it exists, instead of looping over the queue and checking each?
- transport/websocket.rb: connect: Instead of emitting `error`, emit `close` when the connection is closed?
- transport/websocket.rb: connect: Add docu
- transport/websocket.rb: send_transaction: Rename `p` to `promise`
- transport/websocket.rb: close: Remove in favor of using `disconnect`? Or is there a use case for `close`?
- transport/websocket.rb: transaction_id_new: Is this something used throughout? If so, move to base transport?
- transport/websocket.rb: _client: Rename to _create_client
- transport/websocket.rb: client: Docu is off. Maybe remove this alltogether?
- transport/websocket.rb: _timeout_error: Inline? Only 1 usage
- transport/websocket.rb: _promise_wait_timeout: Rename to `_transaction_timeout`, because "promise" is rather an implementation detail. Maybe move this to a constant?

**Comment by kris-lab** [comment]
> Why do we need _self.client - is there another client on the Client?

It is due to `eventemitter` e.g. `client.on :open do`. There is not more current current context available and `self` refers to another one!

**Comment by njam** [comment]
- resource.rb: initialize: add docu
- resource.rb: initialize: remove `id` - not used by any resource?
- resource.rb: How about *always* keeping a reference to the client on all resources? I think it's always needed. We could pass it to the constructor?
- resource/plugin.rb: initialze: Add docu
- resource/plugin.rb: create: Rename `p` to `promise`
- resource/plugin.rb: create: Prefer single over double quotes
- resource/plugin.rb: destroy: Rename `p` to `promise`
- resource/plugin.rb: destroy: Calling `_on_destroyed`, but function is `on_destroyed`
- resource/plugin.rb: on_created / on_destroyed: Make them private?
- resource/plugin.rb: on_destroyed: remove `data`
- resource/plugin.rb: session: Fix docu
- resource/session.rb: initiailize: Add docu
- resource/session.rb: name: I guess it can return `nil` if not created?
- resource/session.rb: create: p -> promise
- resource/session.rb: create: single quotes
- resource/session.rb: on_created/on_destroy: Make private?
- resource/session.rb: on_created: Add docu
- resource/session.rb: on_destroy: Remove data?
- resource/session.rb: heartbeat: single quotes
- error.rb: initialize: Add docu
- client.rb: transport_client: Rename to just `transport`
- client.rb: initialize: Add docu
- client.rb: send: Fix docu (hash instead of string?)
- transport.rb: Streamline send() and send_transaction(): Always expect Hash instead of String?
 - client.rb: ditto
- client.rb: destroy: rename to disconnect
- mountpoint.rb: I'll review it later

**Comment by kris-lab** [comment]
>  transport/websocket.rb: connect: In the error handler we now raise, where will this be rescued? Should it eventually be logged? Maybe add a todo there to remember?

should be handled by parent usage?

>  README: Rtpbrodcast: Last code example: Move .then on new line, so there's only one level of indentation?

Not sure what you mean exactly?

The rest of TRANSPORT is fixed. Will continue on other topics...

**Comment by kris-lab** [comment]
>  resource.rb: initialize: remove id - not used by any resource?

- [ ] it is used in many places?

> resource.rb: How about always keeping a reference to the client on all resources? I think it's always needed. We could pass it to the constructor?

- [ ] not sure! is it really needed in your opinion?

**Comment by kris-lab** [comment]
@njam please review

**Comment by kris-lab** [comment]
@njam replaced `eventemitter` with `events`. please review

**Comment by njam** [comment]
- readme: I meant can we do something like this (untested)?
```ruby
  JanusGateway::Resource::Session.new(client).create.then do |session|
    JanusGateway::Resource::Plugin.new(session, JanusGateway::Plugin::Rtpbroadcast.plugin_name).create
  end.then do |plugin|
    JanusGateway::Plugin::Rtpbroadcast::Resource::Mountpoint.new(plugin, 'test-mountpoint').create
  end.then do |mountpoint|
    # do something with mountpoint
  end
```
- transport.rb: initialize: remove abstract constructor, because other transports might not need url, protocol? We could move this to the websocket transport.
- transport.rb: send: also accept `Hash data`, like for `send_transaction`?
- transport/websocket.rb: initialize: Interesting this `transport`. Do we need to send this value for the janus server to accept the connection?
- transport/websocket.rb: connect: Does this method really ever return?
- transport/websocket.rb: send: Accept `Hash`
- transport/websocket.rb: disconnect: How does `EventMachine.stop` work? Which EMs does it stop? How can we make sure it doesn't stop the wrong EM?
- transport/websocket.rb: ready_state: Unused - remove?
- transport/websocket.rb: is_connected?: No need to compate `nil?` with `false` - just use `!foo.nil? `
- transport/websocket.rb: client: If we have a getter for the client, we should probably use it everywhere (instead of the `@client`?)
- transport/websocket.rb: client: Should it fail if there's no client? Otherwise let's adjust docu
- transport/websocket.rb: _create_client: Fix docu
- transport/websocket_spec.rb: Adjust "describe"
- transport/websocket_spec.rb: Maybe check that the promise is rejected with the correct error?
- transport/websocket_spec.rb: Maybe add some more coverage? :)
- resource.rb: I think it makes sense to have a reference to the client. All resources are somewhat created in reference to a client (or a client's connection), and they might need to send something to the server at some point. Of course some resources can access the client through their parent resources, but nevertheless they're still associated with a client somehow. I think it would be useful to have it. wdyt?
- resource.rb: Add a full getter for `id`, to add some docu (clarifying that it can be `nil`)
- resource/plugin.rb: on_destroyed: Prefix with underscore
- resource/plugin.rb: janus_client: Drop in favor of resource's client getter (tbd)?
- resource/session.rb: name: What's the reason to have this getter? Shouldn't we just use `id`? Can't find usages.
- resource/session.rb: heartbeat: Stop previous heartbeat thread if one running
- resource/session.rb: _on_created: Instead of `send()` can we just call `_on_destroyed`?
- resource/session.rb: _on_created: I don't think we need to pass the `id` when emitting those events, because the events are emitted on a specific session instance (which has the `id` already)?
- resource/session_spec.rb: When testing event callbacks the problem is that we don't check that the callback was actually called. Instead I would create a `Proc`, and expect that it gets executed with:
```ruby
expect(callback).to receive(:call).once.with(something)
```
See for example [here](https://github.com/cargomedia/rndmvr-studio_agent/pull/10/files#diff-fe36555110d9c2c8ec1142f3577d1136R123)
- resource/plugin_spec.rb: Create a spec for the plugin
- rtpbroadcast/resource.rb: Remove?
- rtpbroadcast/resource/mountpoint.rb: No need to include EventEmitter (done in base resource)
- rtpbroadcast/resource/mountpoint.rb: streams: Will fail if not created. How should it be handled?
- rtpbroadcast/resource/mountpoint.rb: destroy: Add docu
- rtpbroadcast/resource/mountpoint.rb: Always use `plugin` getter instead of `@plugin?
- rtpbroadcast/resource/mountpoint.rb: We should also add some assertions (`expect()`). Not sure if the approach with `Proc` is the best, but it should work!

**Comment by kris-lab** [comment]
> readme: I meant can we do something like this (untested)?

- [ ] not working for me :/

> transport/websocket.rb: initialize: Interesting this transport. Do we need to send this value for the janus server to accept the connection?

- [ ] if you mean the `janus-protocol` then yes! we have to do it!

>  transport/websocket.rb: connect: Does this method really ever return?

- [ ] not really. it returns only if EM stops :)

> transport/websocket.rb: disconnect: How does EventMachine.stop work? Which EMs does it stop? How can we make sure it doesn't stop the wrong EM?

- [ ] is it even possible to run multiple EM? This is blocking and doesn't return `self`. Current solution was only one I could find but maybe there us better1?

> transport/websocket.rb: ready_state: Unused - remove?

- [ ] it is used e.g. `is_connected?` ?

> resource/session.rb: _on_created: Instead of send() can we just call _on_destroyed?

no, we cannot call private method from private one!

>  rtpbroadcast/resource.rb: Remove?

- [ ] it is needed otherwise `Plugin::Rtpbroadcast::Resource::Mountpoint` fails with unknown `Plugin::Rtpbroadcast::Resource` !

> rtpbroadcast/resource/mountpoint.rb: streams: Will fail if not created. How should it be handled?

- [ ] should be handled by agent/parent which calls this method? We return `promise` and raise error!

There are still SPEC notes to finish

**Comment by njam** [comment]
> readme

Found the problem - thanks for the help @vogdb - we need to call `.value` before returning in the `then()`. It's strange, and different than in JS, but it works:
JavaScript:
```js
var async = function(value) { return new Promise(function(res) {res(value);}); };

async('hello').then(function(ret) {
  console.log('result1: ' + ret);
  return async(ret + '.');
}).then(function(ret) {
  console.log('result2: ' + ret);
  return async(ret + '.');
}).then(function(ret) {
  console.log('result3: ' + ret);
});
```
Ruby:
```ruby
  async = Proc.new { |value| sleep 1; Concurrent::Promise.new.set(value).execute }

  async.call('hello').then do |value|
    puts "Value: #{value}"
    async.call(value + '.').value
  end.then do |value|
    puts "Value: #{value}"
    async.call(value + '.').value
  end.then do |value|
    puts "Value: #{value}"
  end

  puts "done"
  sleep 5
  puts "done done"
```

> rtpbroadcast/resource/mountpoint.rb: streams: Will fail if not created. -> We return promise and raise error!

Not sure if we talk about the same. I mean this one:
```ruby
    def streams
      @data['data']['stream']['streams']
    end
```
The problem is that `@data` might not be set. There will be a strange error message. Maybe there should be a getter for the data which fails in that case?

**Comment by kris-lab** [comment]
@njam please have a look in meantime

**Comment by njam** [comment]
- transport.rb: ready_state: I would suggest to remove this on the base class, because it doesn't have meaning for all possible transports. Maybe make it private, or even inline on websocket.rb?
- transport/websocket.rb: connect_client: Raise if `@client` already exists?
- transport/websocket.rb: How about renaming `connect -> run` and `connect_client -> connect`?
- transport/websocket_spec.rb: Use `connect` (new)  instead of `run` - and remove the `Thread`
- resource.rb: name: Either return `nil` by default, or change docu to *always* return a string?
- resource/session.rb: initialize: Pass along client to the super call? Or is it done automatically?
- resource/plugin.rb: create: Should we shorten this to `promise.set(self).execute` on one line?
- resource/plugin.rb: destroy: Also shorten-able to just `promise.set(self).execute` I guess (which returns the promise)?
- rtpbroadcast/resource/mountpoint.rb: create: ditto
- rtpbroadcast/resource/mountpoint.rb: destroy: ditto
- plugin.rb: What's the difference of `JanusGateway::Plugin` and `JanusGateway::Resource::Plugin`?  Is the former maybe not needed?
- client.rb: Shouldn't include EventEmitter, because it only proxies `on` calls?
- event_mock.rb: replace in favor of `Faye::WebSocket::API::MessageEvent`

**Comment by kris-lab** [comment]
>  resource/session.rb: initialize: Pass along client to the super call? Or is it done automatically?

by default all `arg` are passed to super. If you need to adjust it then you have to specifically set them.

> plugin.rb: What's the difference of JanusGateway::Plugin and JanusGateway::Resource::Plugin? Is the former maybe not needed?

well we could live without this. Idea was to force each plugin to implement this as that name is handy and fixed! Also to keep proper ruby structure we need `JanusGateway::Plugin` to implement `JanusGateway::Plugin::Rtpbroadcast` - otherwise ruby fails.

> resource.rb: name: Either return nil by default, or change docu to always return a string?

can we always convert to string in getter? wdyt? `@id.to_s`

**Comment by kris-lab** [comment]
@njam please review

**Comment by njam** [comment]
lgtm
